### PR TITLE
Mysql

### PR DESCRIPTION
**Pull request by Redw0od**
Corrected syntax errors, added support to generate connections.xml and launch mysql workbench.

### CODE REVIEW

File: bt/cmd/rds_gui [hunk 1]
```diff
@@ -19,7 +19,7 @@ export BT="${HOME}/.bt"
 aws_defaults
 env_init
 set_team 
-
+exec 3>&1 5>&1
```
Comment (Redw0od) [inline bt/cmd/rds_gui]: added to support file descriptors later in script

File: bt/cmd/rds_gui [hunk 2]
```diff
@@ -50,11 +50,7 @@ default_arg() { if [ -z "$OPTARG" ]; then
     die "No arg for --$OPT option" && usage; fi; 
 }
 
-# lib/utils
-export BT=${HOME}/.bt
-. "${BT}/lib/utils.bash"
-
-while getopts P:d:f:g:h:j:k:o:p:t:u:v:-: OPT; do
+while getopts :P:f:g:j:k:o:p:t:u:v:-:hd OPT; do
```
Comment (Redw0od) [inline bt/cmd/rds_gui]: lines 54 and 55 are repeats of lines 16 and 17.  Fixed syntax errors to enable OPTs processing

File: bt/cmd/rds_gui [hunk 3]
```diff
@@ -82,6 +78,9 @@ while getopts P:d:f:g:h:j:k:o:p:t:u:v:-: OPT; do
 done
 shift $((OPTIND-1)) # remove parsed options and args from $@ list
 
+# display usage if help flag recieved
+if [[ "${usage}" == true ]]; then usage; fi
+
```
Comment (Redw0od) [inline bt/cmd/rds_gui]: Added support to print usage if OPT calls for it

File: bt/cmd/rds_gui [hunk 4]
```diff
@@ -253,13 +252,12 @@ echo "${BT_TOKEN}"
 # launch mysql 
 
 #ssm_keypath+=( devops bt rds "${bt_user}" "${bt_team}" )
-BT_TOKEN
+
```
Comment (Redw0od) [inline bt/cmd/rds_gui]: not sure why this is here, but it generates an error, so i removed it

File: bt/cmd/rds_gui [hunk 5]
```diff
@@ -253,13 +252,12 @@ echo "${BT_TOKEN}"
...
 
 # -----------
 # parse args.
 # -----------
 
-help=false
```
Comment (Redw0od) [inline bt/cmd/rds_gui]: i believe this is superseded by usage above

File: bt/cmd/rds_gui [hunk 6]
```diff
@@ -272,10 +270,6 @@ arg_parse() {
 
 arg_parse
 
-iam_user=restricted  # default
-
-#debug echo instance: ${instance}
-[[ "$help" == true ]] && usage
```
Comment (Redw0od) [inline bt/cmd/rds_gui]: iam_user is redefined 5 lines below.  Help var is never set to true

File: bt/cmd/rds_gui [hunk 7]
```diff
@@ -315,7 +309,7 @@ iam_user=restricted    # default
 # handle tunnel persistence, and LogForwarding capability. 
 
 
-[[ "${BT_CLUSTER}" == "NONE" ]] && { 
+[[ "${bt_cluster}" == "NONE" ]] && {
```
Comment (Redw0od) [inline bt/cmd/rds_gui]: syntax correction

File: bt/cmd/rds_gui [hunk 8]
```diff
@@ -326,7 +320,7 @@ iam_user=restricted    # default
 
 # start the cycle of acquiring a tunnel. 
 # --------------------------------------
-to_debug rds && echo -e "${GREEN}M${NC}: Spinning up a tunnel for ${CLUSTER}". >&3
+to_debug rds && echo -e "${GREEN}M${NC}: Spinning up a tunnel for ${bt_cluster}". >&3
```
Comment (Redw0od) [inline bt/cmd/rds_gui]: syntax correction

File: bt/cmd/rds_gui [hunk 9]
```diff
@@ -347,9 +341,10 @@ to_debug rds echo -e "${CYAN}T${NC}: iam_user is: ${iam_user}" >&3
 #     stdin, stdout, stderr, fd3, fd5
 # 4.) run with nohup
 
-TNL=( ${BT_LAUNCH_DIR}/rds_tunnel "${BT_CLUSTER}" )
+if [[ -z "${BT_LAUNCH_DIR}" ]]; then export BT_LAUNCH_DIR=${0%/*}; fi
+TNL=( ${BT_LAUNCH_DIR}/rds_tunnel_2 "${bt_cluster}" )
```
Comment (Redw0od) [inline bt/cmd/rds_gui:356]: BT_LAUNCH_DIR is never set in this script. Added definition and corrected some syntax

File: bt/cmd/rds_gui [hunk 10]
```diff
@@ -347,9 +341,10 @@ to_debug rds echo -e "${CYAN}T${NC}: iam_user is: ${iam_user}" >&3
...
 echo -ne "\n\n${GREEN}M${NC}: -----------------------------------\n"
-echo -ne     "${GREEN}M${NC}: Spawning tunnel for ${BT_CLUSTER}. \n"
+echo -ne     "${GREEN}M${NC}: Spawning tunnel for ${bt_cluster}. \n"
```
Comment (Redw0od) [inline bt/cmd/rds_gui:361]: corrected syntax

File: bt/cmd/rds_gui [hunk 11]
```diff
@@ -371,21 +366,68 @@ TOKEN="$(aws rds generate-db-auth-token \
          --port 3306                    \
          --username "${iam_user}")"
 
-to_debug rdsc && echo token fingerprint: "${BT_TOKEN:(-7)}"  >&5
+to_debug rdsc && echo token fingerprint: "${TOKEN:(-7)}"  >&5
```
Comment (Redw0od) [inline bt/cmd/rds_gui]: TOKEN is more useful in this debug line

File: bt/cmd/rds_gui [hunk 12]
```diff
@@ -371,21 +366,68 @@ TOKEN="$(aws rds generate-db-auth-token \
...
 
 
 # write a temp file for mysql client config. 
 # -----------------------------------------------
 { 
   # for global opts.
   echo -ne "\n[client]\n"
-  echo -ne "user=\"${iam_user}\"\npassword=\"${BT_TOKEN}\"\n"
+  echo -ne "user=\"${iam_user}\"\npassword=\"${TOKEN}\"\n"
```
Comment (Redw0od) [inline bt/cmd/rds_gui:392]: TOKEN holds the rds auth token

File: bt/cmd/rds_gui [hunk 13]
```diff
@@ -371,21 +366,68 @@ TOKEN="$(aws rds generate-db-auth-token \
...
+        <value type="string" key="sslKey"></value>
+        <value type="int" key="useSSL">3</value>
+        <value type="string" key="userName">${iam_user}</value>
+      </value>
+      <value type="string" key="name">${bt_cluster}</value>
+      <link type="object" struct-name="GrtObject" key="owner">CF52BA91-1D92-4B8C-87CE-54CBD2EACB6A</link>
+    </value>
+  </value>
+</data>
+EOF
```
Comment (Redw0od) [inline bt/cmd/rds_gui]: Encode token for XML characters and generate connection.xml for MySqlWorkbench

File: bt/cmd/rds_gui [hunk 14]
```diff
@@ -395,7 +437,8 @@ arg_parse() {
   declare -a mysql_args
   cfg_args+=( --defaults-extra-file="${tmppath}.cfg" )
   source <( reprocess_mysql_args ${mysql_args[@]} )
-} arg_parse # ------------------------ # build the mysql command
+} 
+arg_parse # ------------------------ # build the mysql command
```
Comment (Redw0od) [inline bt/cmd/rds_gui]: corrected syntax

File: bt/cmd/rds_gui [hunk 15]
```diff
@@ -411,6 +454,9 @@ repoll "${bt_port}" && sleep 0.1 && echo -ne "\n\n"
 to_debug rdsc && echo -e ${PURPLE}C${NC}: command: "${CFG[@]}"  >&5
 #loader_addpath /usr/local/Cellar/mysql/8.0.29/bin
 #loader_flag mysql
-mysql --defaults-extra-file="${HOME}/tmp/rds.cfg"
+#mysql --defaults-extra-file="${HOME}/tmp/rds.cfg"
+#echo "open -n /Applications/MySqlWorkbench.app --args  --configdir ${rds_configdir} --query ${bt_cluster}"
+open -n /Applications/MySqlWorkbench.app --args  --configdir ${rds_configdir} --query ${bt_cluster}
+#echo -n "${TOKEN}" | pbcopy
```
Comment (Redw0od) [inline bt/cmd/rds_gui]: launch MySqlWorkbench with configuration and connection name

File: bt/cmd/rds_tunnel [hunk 16]
```diff
@@ -4,6 +4,7 @@
 # parse rds args
 # ----------------------------
 
+exec 3>&1 5>&1             # redirect 3 and 5 to the pipeline;
```
Comment (Redw0od) [inline bt/cmd/rds_tunnel]: added file descriptors as they're called later in script

File: bt/cmd/rds_tunnel_2 [hunk 17]
```diff
@@ -84,7 +84,7 @@ TNL=( aws ssm start-session )
 TNL+=( --target "${bt_instance}" ) 
 TNL+=( --document-name AWS-StartPortForwardingSessionToRemoteHost )
 TNL+=( --parameters )
-TNL+=( '{"portNumber":["3306"],"localPortNumber":["${bt_port}"],"host":["${bt_endpoint}:3306"]}' )
+TNL+=( '{"portNumber":["3306"],"localPortNumber":["'${bt_port}'"],"host":["'${bt_endpoint}':3306"]}' )
```
Comment (Redw0od) [inline bt/cmd/rds_tunnel_2]: Added quotes to resolve connection errors

File: bt/cmd/rds_gui (hunk 2 above)
Comment (gangofnuns) [reply bt/cmd/rds_gui]: beautiful !   Thanks for noticing this rather subtle, but pernicious flaw.

File: bt/cmd/rds_gui (hunk 6 above)
Comment (gangofnuns) [reply bt/cmd/rds_gui]: Great catch!   Thanks again for this.

**Pull request merged**